zerver/lib/send_email.py

```diff
@@ -148,8 +149,12 @@ def render_templates() -> Tuple[str, str, str]:
     if from_address == FromAddress.support_placeholder:
         from_address = FromAddress.SUPPORT
 
-    # Set the "From" that is displayed separately from the envelope-from
+    # Set the "From" that is displayed separately from the envelope-from.
+    # Amazon SES rejects emails with From names longer than 320 characters (based on RFC limitations);
+    # It's better to send an email without a name than not send it.
     extra_headers["From"] = str(Address(display_name=from_name, addr_spec=from_address))
+    if getsizeof(extra_headers["From"]) > 320:
```

Comment:
Oh, I think I just realized something.  We care about the _encoded_ length of this header.  `getsizeof` is the internal size of the Python object in memory, which is very much not relevant here.  But `len` will only give us the number of codepoints, and `len(....encode("UTF-8"))` will give plain UTF-8 encoding -- but we need the [RFC 2047](https://tools.ietf.org/html/rfc2047.html) encoded length, I think?

```
from django.core.mail.message import sanitize_address

sanitize_address(str(Address(display_name="ȩ̷͍͙̠͂̽͒̊͌͒̽͋̈͆͋͂͝ͅẍ̸̅͌̾͊͒̂͊́̇̈́͑͒́̚͘͝͠a̷͊̈̽̌͑̈́̋̍̀̌̽̾̋͊̕̚͠͝m̴̢̢̨̙̰̙̺̠̭͎̮͓͊̔̐̕͝ͅp̷̛͈͙̫̫̅̓͊̓͑̈͌͌̾̾̚͘͘l̸̂́̏̋̌̀̆̐̂́̐̄̏͒͐̈̈͘ẽ̸̛̜̜͐͆̓͛̃̉̈͛̓̿̅̅̑̕", addr_spec="alex@example.com")), "utf-8")

# '=?utf-8?b?Zcy3zYLMvc2SzIrNjM2SzL3Ni8yIzYbNi82CzZ3Nhc2NzKfNmcygeMy4zIjMhcya?=\n =?utf-8?b?zYzNncy+zYrNksyCzYrNoM2BzIfNhM2RzZjNksyBYcy3zYrMiMy9zJXNoMyMzJo=?=\n =?utf-8?b?zZHNhM2dzIvMjc2AzIzMvcy+zIvNim3MtM2dzYrMlMyVzJDMmcywzJnMos2FzLo=?=\n =?utf-8?b?zKLMoMytzKjNjsyuzZNwzLfMhcyTzYrMms2DzZjMm82RzIjNmM2MzYzMvsy+zYg=?=\n =?utf-8?b?zZnMq8yrbMy4zILMgcyPzIvMjM2AzIbMkMyCzYHMkMyEzI/Nks2YzZDMiMyIZcy4?=\n =?utf-8?b?zIPNkM2GzYPMm82bzIPMicyIzZvMk8yVzL/MhcyFzJHMnMyc?= <alex@example.com>'
```

Reply:
Indeed `getsizeof` is definitely not giving what we want. `len` doesn't seem to work as well but I don't get the difference between `len(string.encode('utf-8'))` and the RFC2047 encoded length ? I must say I am not an expert in codecs but shouldn't it be sufficient to use `len(string.encode('utf-8'))` ? Small test here where I generated a Zalgo Text from https://lingojam.com/ZalgoText with the minimum craziness level and just the letter "a".

![image](https://user-images.githubusercontent.com/37901973/113401585-e40de400-93a3-11eb-9be3-152579bd06d6.png)
![image](https://user-images.githubusercontent.com/37901973/113401612-f0923c80-93a3-11eb-85d5-e830052cfc18.png)

Reply:
> shouldn't it be sufficient to use len(string.encode('utf-8'))

No, because email headers are only assumed to be 7-bit-ASCII safe.  And email address that _renders_ as `a̴` is indeed 3 bytes in utf-8 encoding, but over the wire the header looks like:
```
From: =?utf-8?b?Ycy0?= <alex@example.com>
```
Which is more than 3 bytes.

Reply:
Ok I get it. Then shouldn't we just do
```
extra_headers["From"] = sanitize_address(str(Address(display_name="a̴", addr_spec="alex@example.com")), "utf-8")
if(len(extra_headers["From"]) > 320):
       extra_headers["From"] = sanitize_address(str(Address(addr_spec="alex@example.com")), "utf-8")
```
as sanitize_address gives us the output we want and already in ASCII thus the `len` built-in function should give us the correct size ?

Reply:
Close, but not quite.  We do want to _check_ `len(sanitize_address(str(Address(...)))`, but we don't want to _set_ `extra_headers["From"]` to that -- it already does its own `sanitize_address` behind the scenes, and would likely error because the "sanitized" version (which is a bad name for the function -- it encodes, not sanitizes) contains newlines ([see the also-misnamed `forbid_multi_line_headers`](https://github.com/django/django/blob/main/django/core/mail/message.py#L59-L60), which does that but also encodings)

So something like:
```
extra_headers["From"] = str(Address(display_name=from_name, addr_spec=from_address))
if (len(sanitize_address(extra_headers["From"])) > 320):
    extra_headers["From"] = str(Address(addr_spec=from_address))
```

Reply:
This seems better indeed. I will update my pull request tomorrow to include these changes. Thank you for your helpful clarifications !